File: app/src/main/java/com/checkin/app/checkin/Account/AccountModel.java
package com.checkin.app.checkin.Account;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.List;

@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
public class AccountModel {
    @JsonProperty("pk")
    private String id;

    @JsonProperty("target_pk")
    private String targetPk;

    private ACCOUNT_TYPE accountType;

    private String pic;

    private String name;

    private ObjectNode detail;

    public enum ACCOUNT_TYPE {
        USER(201),
        SHOP_OWNER(202), SHOP_ADMIN(203),
        RESTAURANT_MANAGER(204), RESTAURANT_WAITER(205), RESTAURANT_COOK(206);

        public final int id;
        ACCOUNT_TYPE(int id) {
            this.id = id;
        }

        public static ACCOUNT_TYPE getById(int id) {
            for (ACCOUNT_TYPE type: ACCOUNT_TYPE.values()) {
                if (type.id == id)
                    return type;
            }
            return USER;
        }
    }

    AccountModel() {}

    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public ACCOUNT_TYPE getAccountType() {
        return accountType;
    }

    public String formatAccountType() {
        switch (accountType) {
            case USER:
                return  "User's account";
            case SHOP_OWNER:
                return "Owner's account";
            case SHOP_ADMIN:
                return "Admin's account";
            case RESTAURANT_MANAGER:
                return "Manager's account";
            case RESTAURANT_WAITER:
                return "Waiter's account";
            case RESTAURANT_COOK:
                return "Cook's account";
        }
        return "UNKNOWN ACCOUNT";
    }

    public String getPic() {
        return pic;
    }

    public ObjectNode getDetail() {
        return detail;
    }

    @JsonProperty("acc_type")
    public void setAccountType(int accType) {
        this.accountType = ACCOUNT_TYPE.getById(accType);
    }

    public String getTargetPk() {
        return targetPk;
    }

    @Nullable
    public static AccountModel getByAccountType(@NonNull List<AccountModel> accounts, ACCOUNT_TYPE accountType) {
        for (AccountModel item: accounts) {
            if (item.getAccountType() == accountType)
                return item;
        }
        return null;
    }

    @Override
    public String toString() {
        return getName();
    }
}

File: app/src/main/java/com/checkin/app/checkin/accounts/AccountModel.kt
package com.checkin.app.checkin.accounts

import com.checkin.app.checkin.utility.*
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import com.fasterxml.jackson.databind.node.ObjectNode
import io.objectbox.annotation.Convert
import io.objectbox.annotation.Entity
import io.objectbox.annotation.Id
import io.objectbox.annotation.Transient

@Entity
data class AccountModel(
        @Id var id: Long,
        val pk: Long,
        @JsonProperty("target_pk") val targetPk: Long,
        @JsonProperty("acc_type")
        @JsonDeserialize(using = ACCOUNT_TYPE.Companion.Deserializer::class)
        @JsonSerialize(using = ACCOUNT_TYPE.Companion.Serializer::class)
        @Convert(converter = ACCOUNT_TYPE.Companion.Converter::class, dbType = Int::class)
        val accountType: ACCOUNT_TYPE,
        val pic: String?,
        val name: String,
        @Transient val detail: ObjectNode?
) {

    val formatAccountType: String = when (accountType) {
        ACCOUNT_TYPE.USER -> "User's account"
        ACCOUNT_TYPE.SHOP_OWNER -> "Owner's account"
        ACCOUNT_TYPE.SHOP_ADMIN -> "Admin's account"
        ACCOUNT_TYPE.RESTAURANT_MANAGER -> "Manager's account"
        ACCOUNT_TYPE.RESTAURANT_WAITER -> "Waiter's account"
        ACCOUNT_TYPE.RESTAURANT_COOK -> "Cook's account"
    }

    override fun toString(): String = name
}

enum class ACCOUNT_TYPE(override val value: Int) : EnumIntType {
    USER(201), SHOP_OWNER(202), SHOP_ADMIN(203),
    RESTAURANT_MANAGER(204), RESTAURANT_WAITER(205), RESTAURANT_COOK(206);

    companion object : EnumIntGetter<ACCOUNT_TYPE>() {
        override fun getByValue(value: Int): ACCOUNT_TYPE = EnumIntType.getByValue(value)

        class Deserializer : EnumDeserializer<ACCOUNT_TYPE, Int>(this)
        class Serializer : EnumSerializer<ACCOUNT_TYPE, Int>(this)
        class Converter : EnumConverter<ACCOUNT_TYPE, Int>(this)
    }
}